UART Base Functionality

brainyshadow opened the pull request.

brainyshadow committed:
Moved to stage branch

brainyshadow committed:
Changed rvsw version

brainyshadow committed:
Created queues and threads for UART class

JZJisawesome committed:
Add a test program to test 16550 UART writes!

This test may fail at the moment/hang since the IER register isn't
implemented. But as accuracy improves we should begin passing the test!

Also update copyright info to add Seb. Welcome to the project! :)

brainyshadow committed:
Added MVP writting functionality

brainyshadow committed:
Fixed compile issues

brainyshadow committed:
Fixed compile issues

brainyshadow committed:
Fixed compile issues

brainyshadow committed:
Fixed conditioanl variable issues

brainyshadow committed:
MVP for receieve thread. Need to impliment file descriptor instead of iostream

JZJisawesome committed:
Update rvsw version to bring in new UART tests

Also minor fix for the LSR register!

JZJisawesome committed:
Bring back in minor fix lost in rebase

brainyshadow committed:
Moved receive logic to main thread

brainyshadow committed:
Removed select where it wasn't needed

brainyshadow committed:
Removed select import

nickchan2 commented on lib/uart.h:
```diff
@@ -100,7 +109,14 @@ class Uart {
     uint8_t m_dlm;//Divisor Latch MSB
     uint8_t m_psd;//Prescaler Division
```
Would make more sense if these were grouped in a struct. Also is it intentional that the value of all of these regs is undefined when `Uart` is constructed?

JZJisawesome replied:
I don't know if they need to be grouped in a struct since they are members of the class. Not dissimilar from what we're doing with the `Csr` class I think.

However it would be a good idea to init these to known values, or to something random by using the `irve_fuzzish_rand()` macro in `fuzzish.h` when register contents are unknown in hardware on reset.

This is a pretty easy fix, and I can help if needed. :)

nickchan2 replied:
I think it's okay in `Csr` since the only members of that class are the regs themselves. Here, we have other members besides regs, so grouping them makes it clearer what they are for. Also, that will allow for things like random initializing the whole struct rather than having to do each reg individually. It's a simple fix, all that would need to be done would be wrapping the regs in something like:
```
struct {

    ...

} regs;

JZJisawesome replied:
Ah I see that makes sense. No strong preferences on my end :)

nickchan2 commented on lib/uart.cpp:
```diff
@@ -56,16 +61,20 @@ using namespace irve::internal;
 
 Uart::Uart() {
-    //TODO
+    transmit_thread = std::thread(&Uart::transmit_thread_function, this);
```
Add `this->` in front of `transmit_thread`

brainyshadow replied:
Thank you. I've addressed this in my most recent commit.

nickchan2 commented on lib/uart.cpp:
```diff
@@ -76,7 +85,14 @@ uint8_t Uart::read(Uart::Address register_address) {
+                    data = receive_queue.front();
+                    receive_queue.pop();
+                }else{
+                    assert(false && "Tried to read from empty input");
```
Is this assertion just for RISCV software debugging? If so, add a comment to clarify, since I don't think it's ever an emulator bug to read from this register

JZJisawesome replied:
I'm cool with this assert failing if software reads from the register since it will be good at helping us catch software / kernel bugs.

A comment clarifying this wouldn't hurt. Or perhaps change the message to "Buggy software tried to read from empty input" to accomplish a similar effect. :)

nickchan2 replied:
Yeah, would def be good to clarify since from my understanding, we've been using assertions to find bugs in the emulator, not in software running on it

JZJisawesome replied:
Agreed, and then once the kernel is working and RVSW stuff is more stabilized, we can remove it and probably just return 0.

brainyshadow replied:
Good point. I've added a comment in my most recent commit.

nickchan2 commented on lib/uart.cpp:
```diff
@@ -100,9 +116,8 @@ uint8_t Uart::read(Uart::Address register_address) {
-            assert(false && "TODO");//TODO
-            break;
+        case Uart::Address::LSR: {
+            return this->m_lsr | (1 << 5);//We are always ready to transmit
```
Is 5 the position of the ready to transmit bit? Maybe add `constexpr uint32_t LSR_TX_READY_POS = 5U;` above this line and use it in the bit shift to avoid a magic number. Also, when bit shifting, it's better practice to use unsigned values (change `1` to `1U`)

JZJisawesome replied:
Yeah this is my bad, I added that in one of my commits to the branch. Sounds like a good fix to me :)

nickchan2 commented on lib/uart.cpp:
```diff
@@ -187,10 +181,34 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
+            receive_queue.push(data);
+            this->m_isr |= 0b1;
+    }else if(receive_queue.size() == 0){
+        this->m_isr &= 0b11111110;
```
Use `&= ~` when clearing bits (change to `&= ~0b1`)

JZJisawesome replied:
Kind of a stylistic thing, honestly I don't mind either way. Whatever you two think is best :)

nickchan2 replied:
Yeah obviously it doesn't matter that much. Reason I make the suggestion is because:
- In my experience, the convention for bit clearing is always `&= ~`.
- The current code won't work if the width of `m_isr` changes. Not that it ever would but I'd always rule in favor of good practice

JZJisawesome replied:
Oh cool I didn't think about the size changing in the future. That's why people do that, I always wondered!

brainyshadow replied:
Thank you. I've addressed this in my most recent commit.

nickchan2 commented on lib/uart.cpp:
```diff
@@ -124,31 +138,11 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
+                    std::lock_guard<std::mutex> lock(this->transmit_mutex); 
+                    this->transmit_condition_variable.notify_one();          
+                }       
+                this->async_transmit_queue.push(data);
```
Why do we push to the queue after waking up the thread rather than before?

JZJisawesome replied:
Good point. Thankfully it's a simple fix I think! Just gotta move the push to before that scope block for the lock :)

brainyshadow replied:
This was silly. My bad. Thank you. Should be addressed in my most recent commit.

JZJisawesome commented:
Just a couple of minor tweaks and we should be good to go! I'm happy to help if desired (especially with mistakes introduced in my commits).

brainyshadow committed:
Addressed PR comments

brainyshadow committed:
Changed inits to random values

brainyshadow committed:
Added cast

nickchan2 commented on lib/uart.cpp:
```diff
@@ -56,16 +61,21 @@ using namespace irve::internal;
 
 Uart::Uart() {
-    //TODO
+    this->regs = {static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand()), static_cast<uint8_t>(irve_fuzzish_rand())};
```
Now that the regs are grouped in a struct, you could do this instead:

    irve_fuzzish_meminit(&this->regs, sizeof(this->regs));

brainyshadow replied:
Fair enough. I personally think the current way is more clear but if you think `irve_fuzzish_meminit(&this->regs, sizeof(this->regs));` is better, we can change this.

nickchan2 replied:
Yeah I see where you're coming from, I'm good with keeping it as is

nickchan2 commented on lib/uart.cpp:
```diff
@@ -174,23 +170,47 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
-    return this->m_isr & 0b1;
+void Uart::update_receive() {
+    uint8_t data;
+    ssize_t bytesRead = ::read(this->receive_file_fd, &data, 1);
```
Might be a good idea to handle a read error (could just assert false)

nickchan2 commented on lib/uart.cpp:
```diff
@@ -174,23 +170,47 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
+        while(this->async_transmit_queue.size() > 0){
+            char data = char(this->async_transmit_queue.front());
+            this->async_transmit_queue.pop();
+            std::cout<<data<<std::flush;
```
I think our original plan was to use a fd for both input and output

nickchan2 commented on lib/uart.cpp:
```diff
@@ -174,23 +170,47 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
+        this->transmit_condition_variable.wait(lock);
+        while(this->async_transmit_queue.size() > 0){
+            char data = char(this->async_transmit_queue.front());
+            this->async_transmit_queue.pop();
```
In general this could be a synchronization error. Reading the front and popping from the queue should happen in the same lock. Otherwise, after the read and before the pop, another thread could change the queue. In our case, I don't think this would ever be a problem, and I realize that right now our thread safe queue doesn't implement `pop_front` which is what we should be using here.

brainyshadow replied:
This is a good point. I can try and implement a pop_front to our threadsafe queue.

brainyshadow committed:
Fixed bit shifting mistake

JZJisawesome committed:
Merge branch 'staging' into personal/seb/primary

Also fix conflicts to do with copyright additions :)

JZJisawesome commented:
Hey Seb I did some testing and while the `uart_write_test` seems to work, `uart_read_test` seems to be a bit broken, either refusing keyboard input or just printing `You just sent the character: '�'!` repeatedly (where that character it is sending is corrupted).

This may be a bug with my test program, or with the async read logic. I'll look into it. If it's a simple fix I'll commit it and merge tonight :)

JZJisawesome committed:
Oops my bad, bring in support elsewhere in the emulator needed to handle UART interrupts

This was breaking Seb's logic since we were never calling his interrupt
pending function, and thus his update function that read characters
asynchronously was never being called either. This fixes that.

Also implements the Data Ready bit in the LSR in a similar way to the
one in the ISR with the pending bit. Fixes my uart test which assumed
both were implemented :)

JZJisawesome commented:
Yep my bad. The test was okay but I forgot to actually call the interrupt_pending function you wrote. I did to work to call it but it was in one of my branches (I never committed it to yours). My latest commit should fix things.
